src/Symfony/Component/Security/Http/AccessToken/OidcAccessTokenHandler.php

```diff
@@ -0,0 +1,60 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Security\Http\AccessToken;
+
+use Psr\Log\LoggerInterface;
+use Symfony\Component\Security\Core\Exception\BadCredentialsException;
+use Symfony\Component\Security\Core\Exception\InvalidOidcUserException;
+use Symfony\Contracts\HttpClient\HttpClientInterface;
+
+/**
+ * The token handler validates and retrieves the user identifier from an OIDC server using the token.
+ *
+ * @author Vincent Chalamon <vincentchalamon@gmail.com>
+ */
+final class OidcAccessTokenHandler implements AccessTokenHandlerInterface
+{
+    private HttpClientInterface $oidcClient;
+    private ?LoggerInterface $logger;
+    private string $userProperty;
+
+    public function __construct(HttpClientInterface $oidcClient, LoggerInterface $logger = null, string $userProperty = 'email')
+    {
+        $this->oidcClient = $oidcClient;
+        $this->logger = $logger;
+        $this->userProperty = $userProperty;
+    }
+
+    public function getUserIdentifierFrom(string $accessToken): string
+    {
+        try {
+            // Call the OIDC server to retrieve the user info
+            // If the token is invalid or expired, the OIDC server will return an error
+            $userinfo = $this->oidcClient->request('GET', 'protocol/openid-connect/userinfo', [
```

Comment:
The URL could be a parameter so that we can hit external services.

Reply:
In addition and to avoid hitting the external service each time the handler is called, we can leverage on the cache system.

Reply:
On what can we rely to expire this cache? I'm in favor of calling the external service each time needed. We don't know when data could change about a user.

Reply:
The user identifier is not suppose to change. In particular, the `sub` claim (if used) should always be the same (https://openid.net/specs/openid-connect-core-1_0.html#SubjectIDTypes), so you cannot expect to receive a different value for a given access token. If you use the email address, I concur with you and it could change, but not that often. Other claims are not relevant here. 

For me, it is safe to cache the response for the lifetime of the access token.
But as this lifetime is not known, it could be an arbitrary, but customizable, value to adapt depending on the use case.

Reply:
> you cannot expect to receive a different value for a given access token

The OIDC server can invalidate the token thanks to the `jti` key in case the token has been stolen or if the password has been changed. I think that it's safer always to call it. We could cache the token if the response contains standard cache headers (that would mean that the server doesn't have this kind of features), but it will introduce some complexity that is probably not worth it.

Reply:
> The OIDC server can invalidate the token thanks to the jti key in case the token has been stolen or if the password has been changed.
This is correct. Note that, even if the access token is invalidated, the user identifier is not supposed to change.
Also, later in the authentication process, the UserProvider will hit the userinfo endpoint to get user details and return the actual User object with all the details.
As you mentioned, caching can be done directly on the HttpClient side and not here.

Reply:
> Also, later in the authentication process, the UserProvider will hit the userinfo endpoint to get user details and return the actual User object with all the details.

Not necessarily. In some use-cases, the user details can be in a local database, for example to store more informations than the OIDC server could.

I'm in favor of no cache on the HTTP client to keep this system open to project behaviors but still secure in case of token invalidation on OIDC server side.

Reply:
The cache can be handled in the injected HTTP client anyway, so no need to add caching logic here.